### PR TITLE
Feature get key

### LINKED ISSUE #28: Expose a function to get and cache a matching JWSK key
Expose a function to get a JWSK (no just secret), and save it to cache

### PR DESCRIPTION
**Pull request by conor909**
Closes #28 

Allows better support for JWKS for fastify-jwt
https://github.com/fastify/fastify-jwt/issues/132#issue-808853440

### CODE REVIEW

File: src/get-jwks.js [hunk 1]
```diff
@@ -4,63 +4,66 @@ const fetch = require('node-fetch')
...
   const max = cacheProps.max || 100
   const ttl = cacheProps.ttl || 60 * 1000
   const cache = lru(max, ttl)
 
-  async function getSecret (signatures) {
+  async function getPublicKey (signatures) {
     const { domain, alg, kid } = signatures
     const cacheKey = `${alg}:${kid}:${domain}`
-    const cachedSecret = cache.get(cacheKey)
+    const cachedJwk = cache.get(cacheKey)
```
Comment (simoneb) [inline src/get-jwks.js:18]: all this function should do is call getJwk and convert the result to PEM

**Review by simoneb — CHANGES_REQUESTED** [review]

File: src/get-jwks.js (hunk 1 above)
Comment (conor909) [reply src/get-jwks.js]: Ah I see, yep thanks!

File: test/getJwk.spec.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
+'use strict'
+
+const nock = require('nock')
+const t = require('tap')
+
+const buildGetJwks = require('../src/get-jwks')
+
+t.beforeEach((done, t) => {
```
Comment (simoneb) [inline test/getJwk.spec.js:8]: the t argument in the callback is unused, same at line 13

File: test/getJwk.spec.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+t.test('getJwk should return an error if the request fails', async t => {
+  nock('https://localhost/').get('/.well-known/jwks.json').reply(500, { msg: 'no good' })
+  try {
+    const getJwks = buildGetJwks()
+    await getJwks.getJwk({ domain: 'https://localhost/', alg: 'ALG', kid: 'SOME_KEY' })
+  } catch (e) {
+    t.equal(e.message, 'Internal Server Error')
+    t.same(e.body, { msg: 'no good' })
+  }
```
Comment (simoneb) [inline test/getJwk.spec.js:27]: what happens if no error is thrown in this test?

File: test/getJwk.spec.js (hunk 2 above)
Comment (conor909) [reply test/getJwk.spec.js]: Thanks for spotting that

File: test/getJwk.spec.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  const alg = localKey.alg
+  const kid = localKey.kid
+
+  getJwks.cache.set(`${alg}:${kid}:${domain}`, localKey)
+  try {
+    const secret = await getJwks.getJwk({ domain, alg, kid })
+    t.ok(secret)
+  } catch (e) {
+    t.throws(e)
+  }
```
Comment (simoneb) [inline test/getJwk.spec.js:134]: I don't think the catch clause is necessary here

File: test/cache.spec.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  const domain = 'https://localhost/'
+  const localKey = jwks.keys[0]
+  const alg = localKey.alg
+  const kid = localKey.kid
+
+  getJwks.cache.set(`${alg}:${kid}:${domain}`, localKey)
+
+  try {
+    const secret = await getJwks.getPublicKey({ domain, alg, kid })
+    const key = await getJwks.getPublicKey({ domain, alg, kid })
```
Comment (simoneb) [inline test/cache.spec.js:39]: you're calling the same function here

File: test/cache.spec.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    t.ok(secret)
+    t.ok(key)
+  } catch (e) {
+    t.throws(e)
+  }
```
Comment (simoneb) [inline test/cache.spec.js:44]: I don't think the exception needs to be caught here, are you expecting one in the first place?

File: test/cache.spec.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+t.test('if initialized without any cache settings it should use default values', async t => {
+  const getJwks = buildGetJwks()
+  const domain = 'https://localhost/'
+  const cache = getJwks.cache
+  const localKey = jwks.keys[0]
+  const alg = localKey.alg
+  const kid = localKey.kid
+  cache.set(`${alg}:${kid}:${domain}`, localKey)
+  const secret = await getJwks.getPublicKey({ domain, alg, kid })
```
Comment (simoneb) [inline test/cache.spec.js:55]: please rename variables to match the function name. "secret" should not exist anymore anywhere in the codebase

File: test/cache.spec.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  const alg = localKey.alg
+  const kid = localKey.kid
+  const cache = getJwks.cache
+  cache.set(`${alg}:${kid}:${domain}`, localKey)
+  const secret = await getJwks.getPublicKey({ domain, alg, kid })
+  const key = await getJwks.getJwk({ domain, alg, kid })
+  t.ok(secret)
+  t.ok(key)
+  t.equal(cache.get(`${alg}:${kid}:${domain}`), localKey)
+  await getJwks.clearCache()
```
Comment (simoneb) [inline test/cache.spec.js:77]: clearCache is not async, no need to await it

**Review by simoneb — CHANGES_REQUESTED** [review]
The pattern you're using for writing tests that expect errors does not work. If no error is thrown, those tests will all succeed. Before changing it sync with me to explain how you intend to address it please

File: test/cache.spec.js (hunk 5 above)
Comment (conor909) [reply test/cache.spec.js:39]: That was a silly mistake, thanks

File: test/getJwk.spec.js (hunk 3 above)
Comment (conor909) [reply test/getJwk.spec.js:27]: I see!  It's not something I thought about, thank you!

File: test/getJwk.spec.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+t.test('getJwk should return an error if the request fails', async t => {
+  nock('https://localhost/').get('/.well-known/jwks.json').reply(500, { msg: 'no good' })
+  try {
+    const getJwks = buildGetJwks()
+    await getJwks.getJwk({ domain: 'https://localhost/', alg: 'ALG', kid: 'SOME_KEY' })
+  } catch (e) {
+    t.equal(e.message, 'Internal Server Error')
+    t.same(e.body, { msg: 'no good' })
+  }
+  t.throws(() => { throw new Error('An error was thrown') }, {})
```
Comment (simoneb) [inline test/getJwk.spec.js:28]: useless

Comment (conor909) [reply test/getJwk.spec.js]: Thanks for showing that, I see how completely useless that is now!

**Comment by mcollina** [comment]
Can you please add an integration test against fastify-jwt and fast-jwt?

File: test/fastify-integrations.spec.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    {
+      alg: 'HS256',
+      kid: 'KEY_1',
+      e: 'AQAB',
+      kty: 'RSA',
+      n: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.zbgd5BNF1cqQ_prCEqIvBTjSxMS8bDLnJAE_wE-0Cxg',
+      use: 'sig'
+    }
+  ]
+}
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:38]: let's not repeat these, extract them in a module

File: test/fastify-integrations.spec.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  t.test('Fastify should register without missing secret errors', function (t) {
+    const fastify = Fastify()
+    fastify.register(fastifyJwt, {
+      secret: publicKey
+    })
+    fastify.ready((err) => {
+      t.equal(err, undefined)
+      t.done()
+    })
+  })
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:58]: we should also write a test to check that fastify-jwt actually works with jwks. also, this is not the integration we're looking for. we want an integration that makes fastify-jwt invoke our library. the point is, we don't have a jwt token available to extract the key information from, we should let fastify-jwt provide one to the getPublicKey function

File: test/cache.spec.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+t.test('if there is already a key in cache, it should not make a http request', async t => {
+  const getJwks = buildGetJwks()
+  const domain = 'https://localhost/'
+  const localKey = jwks.keys[0]
+  const alg = localKey.alg
+  const kid = localKey.kid
+
+  getJwks.cache.set(`${alg}:${kid}:${domain}`, localKey)
```
Comment (mcollina) [inline test/cache.spec.js:36]: I would prefer these to be actual integration tests, calling a server that exposes the jwks.

File: test/jwks.spec.js [hunk 13]
```diff
@@ -1,235 +0,0 @@
-'use strict'
```
Comment (mcollina) [inline test/jwks.spec.js:1]: why was this file removed?

File: test/getPublicKey.spec.js [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+      e: 'AQAB',
+      kty: 'RSA',
+      n: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvZSBCbG9nZ3MiLCJhZG1pbiI6dHJ1ZSwianRpIjoiZmYxMGYxODUtYWI4MS00OGNhLWFmYjUtN2RjYWEzM2ZjODM1IiwiaWF0IjoxNjE0MTAzOTE2LCJleHAiOjE2MTQxMDc1NDl9.4PQiX1jCDiTbRnwWusQHW2UNHxpGgpcRaUxSt4K6C8I',
+      use: 'sig'
+    }
+  ]
+}
+
+const jwksNoKeys = {}
+const jwksEmptyKeys = { keys: [] }
```
Comment (mcollina) [inline test/getPublicKey.spec.js:142]: why are those here?

File: test/getJwk.spec.js [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+      e: 'AQAB',
+      kty: 'RSA',
+      n: 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvZSBCbG9nZ3MiLCJhZG1pbiI6dHJ1ZSwianRpIjoiZmYxMGYxODUtYWI4MS00OGNhLWFmYjUtN2RjYWEzM2ZjODM1IiwiaWF0IjoxNjE0MTAzOTE2LCJleHAiOjE2MTQxMDc1NDl9.4PQiX1jCDiTbRnwWusQHW2UNHxpGgpcRaUxSt4K6C8I',
+      use: 'sig'
+    }
+  ]
+}
+
+const jwksNoKeys = {}
+const jwksEmptyKeys = { keys: [] }
```
Comment (mcollina) [inline test/getJwk.spec.js:152]: why are those here?

**Review by mcollina — CHANGES_REQUESTED** [review]
It's still missing the fast-jwt tests

File: package.json [hunk 16]
```diff
@@ -36,8 +36,12 @@
     "eslint-plugin-import": "^2.22.1",
     "eslint-plugin-node": "^11.1.0",
     "eslint-plugin-promise": "^4.3.1",
+    "fast-jwt": "^1.1.0",
+    "fastify": "^3.12.0",
+    "fastify-jwt": "^2.3.0",
     "nock": "^13.0.7",
     "node-cache": "^5.1.2",
+    "request-promise": "^4.2.6",
```
Comment (mcollina) [inline package.json:44]: request is deprecated. Remove and don't use. Use node-fetch.

File: constants.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+module.exports = {
```
Comment (mcollina) [inline constants.js:1]: use strict is missing

File: constants.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+        alg: 'HS256',
+        kid: 'KEY_1',
+        e: 'AQAB',
+        kty: 'RSA',
+        n: "123",
+        use: 'sig'
+      }
+    ]
+  }
+}
```
Comment (mcollina) [inline constants.js:22]: This file seems something needed for tests. Should it live in the test folder?

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by simoneb** [comment]
@conor909 To write a working integration test for fastify-jwt you need to create and sign a token with a known private key. See how it's done in fastify-jwt https://github.com/fastify/fastify-jwt/tree/master/certs.

The process would be:

- generate key pair
- use private key to sign token
- use nock to return public key the the JWKS format (Matteo asked to have this as a real running server but I don't mind if we keep nock for the time being)
- provide the token in a request to your fastify server
- see that the token is verified successfully (because the public key returned by nock is used to verify the token, which was signed with the corresponding private key)

File: package.json [hunk 19]
```diff
@@ -7,7 +7,7 @@
     "node": ">=10"
   },
   "scripts": {
-    "test": "tap",
+    "test": "tap --test-ignore=test/constants.js",
```
Comment (simoneb) [inline package.json:10]: this shouldn't be necessary. in any case let's whitelist instead of blacklisting, if it's necessary

File: test/fastify-integrations.spec.js [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    secret: {
+      private: {
+        passphrase: 'mysecret',
+        key: readFileSync(path.join(__dirname, 'private.pem'), 'utf8')
+      },
+      public: async (request, token, callback) => {
+        const { header: { kid, alg }, payload: { iss } } = token
+        const publicKey = await getJwks.getPublicKey({ kid, domain: iss, alg })
+        callback(null, publicKey)
+      }
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:50]: this looks good now, and in fact a scenario I had not considered. I'm not sure it's a perfectly valid one to be honest. if you're using JWKS, you probably have the keys available so you don't need to ask them to an external service. In any case, we can keep this test but we should have one which is the usage you've have when integrating with an external service like cognito. the application doesn't know the private key, it only gets the signed token and uses the public key to verify it. in other words, I believe we need a test which doesn't use the secret.private property because we don't have that available. Instead, it uses the function version of the secret property or simply secret.public (not sure this works if you provide public but not private, please try it out) to provide a function which has the same implementation as your secret.public function above. then, you would not use fastify.jwt.sign to sign the token but you would sign it outside of the app using jsonwebtoken. I hope this is clear enough

Comment (simoneb) [reply test/fastify-integrations.spec.js:50]:
```
const jsonwebtoken = require('...')

const token = jsonwebtoken.sign({ payload }, privateKey, { issuer, keyid, algorithm })
```

File: package.json (hunk 19 above)
Comment (conor909) [reply package.json]: I'm not sure it's 100% necessary, I just didn't like the output in the console telling me that it skipped over these files due to no tests found, so I added the whitelist instead `--files=test/**/*.spec.js`

File: test/fastify-integrations.spec.js [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+      reply.send(err)
+    }
+  })
+  fastify.get('/', async (request, reply) => {
+    return request.user.name
+  })
+
+  t.tearDown(fastify.close.bind(fastify))
+
+  fastify.listen(3000)
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:67]: you don't need to start the server

File: test/fastify-integrations.spec.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+      await request.jwtVerify()
+    } catch (err) {
+      reply.send(err)
+    }
+  })
+  fastify.get('/', async (request, reply) => {
+    return request.user.name
+  })
+
+  t.tearDown(fastify.close.bind(fastify))
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:65]: since you don't need to start the server, you don't need to close it

File: test/fastify-integrations.spec.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+  const getJwks = buildGetJwks()
+
+  fastify.register(require('fastify-jwt'), {
+    ignoreExpiration: true,
+    decode: { complete: true },
+    secret: async (request, token, callback) => {
+      const { header: { kid, alg }, payload: { iss } } = token
+      const publicKey = await getJwks.getPublicKey({ kid, domain: iss, alg })
+      callback(null, publicKey)
+    },
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:50]: can you check if the async version of this works too? if so, let's switch to that. by this I mean, instead of accepting a callback, return the publicKey found

File: test/fastify-integrations.spec.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+    noAuthorizationInHeaderMessage: (err) => { console.log('Test Error: ', err.message); return err.message },
+    authorizationTokenExpiredMessage: (err) => { console.log('Test Error: ', err.message); return err.message },
+    authorizationTokenInvalid: (err) => { console.log('Test Error: ', err.message); return err.message }
+  }
+
+  const fastify = Fastify()
+  const getJwks = buildGetJwks()
+
+  fastify.register(require('fastify-jwt'), {
+    ignoreExpiration: true,
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:44]: since we are signing the token in the test, we don't need to ignore the expiration

File: test/fastify-integrations.spec.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    decode: { complete: true },
+    secret: async (request, token, callback) => {
+      const { header: { kid, alg }, payload: { iss } } = token
+      const publicKey = await getJwks.getPublicKey({ kid, domain: iss, alg })
+      callback(null, publicKey)
+    },
+    messages: customErrorMessages
```
Comment (simoneb) [inline test/fastify-integrations.spec.js:51]: why do we need to override the messages?

**Review by simoneb — CHANGES_REQUESTED** [review]
Looks better, the tests still need some minor tweaks though

File: test/fastify-integrations.spec.js (hunk 24 above)
Comment (conor909) [reply test/fastify-integrations.spec.js]: Yes I meant to remove that, thanks!

File: test/fastify-integrations.spec.js (hunk 22 above)
Comment (conor909) [reply test/fastify-integrations.spec.js]: Do I not need to start the server to test the request?

Comment (simoneb) [reply test/fastify-integrations.spec.js]: nope https://www.fastify.io/docs/latest/Testing/

Comment (conor909) [reply test/fastify-integrations.spec.js]: Ah, I see that I don't now, thanks!

File: test/fastify-integrations.spec.js (hunk 23 above)
Comment (simoneb) [reply test/fastify-integrations.spec.js]: fair enough

File: package-lock.json [hunk 26]
```diff
@@ -1,8 +1,6739 @@
 {
   "name": "get-jwks",
   "version": "2.1.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
```
Comment (mcollina) [inline package-lock.json:4]: please remove this file

File: test/constants.js [hunk 27]
```diff
@@ -0,0 +1,39 @@
+'use strict'
+
+// https://russelldavies.github.io/jwk-creator/
```
Comment (mcollina) [inline test/constants.js:3]: what is this?

File: test/jwks.spec.js (hunk 13 above)
Comment (mcollina) [reply test/jwks.spec.js:1]: Why was this removed?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/jwks.spec.js (hunk 13 above)
Comment (conor909) [reply test/jwks.spec.js:1]: it was just renamed, and broken up into multiple test files

File: test/constants.js (hunk 27 above)
Comment (conor909) [reply test/constants.js:3]: just a tool I found handy for creating jwk's for tests, I went through a few packages that didn't work, so thought I'd leave it there for other devs, but I'll remove it if it's messy

File: README.md [hunk 28]
```diff
@@ -58,4 +72,39 @@ const getJwks = buildGetJwks({
...
+const getJwks = buildGetJwks()
+
+fastify.register(fjwt, {
+  decode: { complete: true },
+  secret: async (request, token, callback) => {
+    const { header: { kid, alg }, payload: { iss } } = token
+    const publicKey = await getJwks.getPublicKey({ kid, domain: iss, alg })
+    callback(null, publicKey)
+  }
+})
```
[the remark this answers is not in the document]
Comment (simoneb) [reply README.md:97]: this is the reason why I asked earlier if this function exposed by fastify-jwt supported promises. Matteo is suggesting to avoid mixing async/await with callback based code, but here we have the fastify-jwt function which doesn't seem to support promises, and the getPublisKey function which only supports promises. I don't think we should implement a callback flavor of getPublicKey, we should go forward in terms of javascript features, not backwards. 

Hence, I don't see a way to avoid mixing promises and callback code, therefore this code is good enough for me, with the exception that in order to make the two work properly together you need to wrap the await statement in try/catch.

Alternatively, instead of using async/await you can use plain promise methods and do something like:

```js
getPublicKey(...).then(publicKey => callback(null, publicKey), callback)
```

File: package-lock.json (hunk 26 above)
Comment (simoneb) [reply package-lock.json:4]: put it in .gitignore basically

File: test/constants.js (hunk 27 above)
Comment (simoneb) [reply test/constants.js:3]: yes let's not reference it here. a better place is the docs but it's somewhere in another package's docs so we don't strictly need it here. remove

**Review by simoneb — APPROVED** [review]
LGTM

File: README.md (hunk 28 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply README.md:95]: Please do not mix promises and callbacks. __Just use async/await__.

File: test/fastify-jwt-integrations.spec.js [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+        .then(publicKey => callback(null, publicKey), callback)
+    }
+  })
+
+  fastify.addHook('onRequest', async (request, reply) => {
+    try {
+      await request.jwtVerify()
+    } catch (err) {
+      reply.send(err)
+    }
```
Comment (mcollina) [inline test/fastify-jwt-integrations.spec.js:42]: you can remove the try-catch

**Review by mcollina — CHANGES_REQUESTED** [review]

File: README.md (hunk 28 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply README.md:95]: I understood this, no worries!

[the remark this answers is not in the document]
Comment (simoneb) [reply README.md:95]: give us an approval when you're happy with these changes :)

**Review by mcollina — APPROVED** [review]
lgtm
